### DISCUSSION TITLE
How to include Iceoryx in Makefile?

### DISCUSSION THREAD

**Discussion by piyushmz** [eclipse-iceoryx/iceoryx#1869] [Q&A]
I'm having a camera driver which is written in C with Makefile having gstreamer files in it which generates an `main.elf` file to run multiple cameras simultaneously. I want to include Iceoryx in a Makefile, is there any Makefile present so that I can take some reference.
In Makefile i've included following
```
%.o: %.c
	$(CC) $(CFLAGS)  $(FLAGS)  -I/usr/include/gstreamer-1.0 -I/usr/local/include/iceoryx/v2.90.0/ -c $< -o $@

%: %.c
	$(CC) $(CFLAGS)  $(FLAGS)  -I/usr/include/gstreamer-1.0 -I/usr/local/include/iceoryx/v2.90.0/ -o $@ $<

```
I get the following error while making 
```
main.c:9:10: fatal error: iceoryx_binding_c/publisher.h: No such file or directory
    9 | #include "iceoryx_binding_c/publisher.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

One more thing Suppose I create  simple file and include iceoryx in it how do I compile it without CMakeLists.txt? I want to compile it as a way like we do `gcc -o main main.c` for output

Thanks

**Comment by elfenpiff** [comment]
@piyushmz I think the include path is incorrect. You have to add at least:
* `-I/usr/local/include/iceoryx/v2.90.0/iceoryx_binding_c/include`
* `-I/usr/local/include/iceoryx/v2.90.0/iceoryx_posh/include`
* `-I/usr/local/include/iceoryx/v2.90.0/iceoryx_hoofs/include`
* `-I/usr/local/include/iceoryx/v2.90.0/iceoryx_platform/linux/include`

But please use a modern build system like cmake or bazel which are officially supported by iceoryx. If you want to use it in a bare metal make file you will encounter now and in the future a lot of maintenance issues where you are on your own.

It will be some effort to use cmake but I really recommend to invest the time here otherwise you will have to invest much more time to get iceoryx and your whole project running whenever we update iceoryx or some other underlying library changes. And there are big refactorings coming this year so that you will most likely face the challenge.

**Comment by piyushmz** [comment]
@elfenpiff Thanks for clarification. I'm on Jetson Orin Ubuntu 20.04
Following is my CMakeLists.txt

```
cmake_minimum_required(VERSION 3.16)
project(e-multicam)

include(GNUInstallDirs)

find_package(iceoryx_platform REQUIRED)
find_package(iceoryx_posh CONFIG REQUIRED)
find_package(iceoryx_hoofs CONFIG REQUIRED)
find_package(iceoryx_binding_c CONFIG REQUIRED)

include(IceoryxPackageHelper)
include(IceoryxPlatform)
include(IceoryxPlatformSettings)

set(CMAKE_C_COMPILER ${CROSS_COMPILE}gcc)

set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -g -Wall -pedantic -DVERSION=\\\"${VERSION}\\\"")

set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -lpthread")

set(CMAKE_CXX_STANDARD 11)

find_package(PkgConfig)
pkg_check_modules(PKG_CONFIG_MODULES gstreamer-1.0 gstreamer-video-1.0 gstreamer-app-1.0 x11)
include_directories(${PKG_CONFIG_MODULES_INCLUDE_DIRS})

file(GLOB_RECURSE SRC_FILES "*.c")

add_executable(e-multicam ${SRC_FILES})
target_link_libraries(e-multicam pthread m jpeg ${PKG_CONFIG_MODULES_LDFLAGS} )

install(TARGETS e-multicam DESTINATION ./install)

```
I've included the icedelivery_in_c publisher in the main code for testing purposes.

Same is the error I get

```
/home/minuszero/Work/e-multicam/main.c:9:10: fatal error: iceoryx_binding_c/publisher.h: No such file or directory
    9 | #include "iceoryx_binding_c/publisher.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Can you please review where I'm going wrong.
Thanks.

**Reply by elfenpiff** [reply]
@piyushmz 

1. Never use `include_directories` but write `target_include_directories(e-multicam ${PKG_CONFIG_MODULES_INCLUDE_DIRS}`
2. (C++) You have to add iceoryx in `target_link_libraries(e-multicam ... iceoryx_posh::iceoryx_posh)`
3. (C) You have to add iceoryx in `target_link_libraries(e-multicam ... iceoryx_binding_c::iceoryx_binding_c)`

Does this work?

**Reply by piyushmz** [reply]
After including `target_include_directories(e-multicam ${PKG_CONFIG_MODULES_INCLUDE_DIRS}`

```
CMake Error at CMakeLists.txt:26 (target_include_directories):
  Cannot specify include directories for target "e-multicam" which is not
  built by this project.
```

**Reply by elfenpiff** [reply]
The `target_include_directories` must come after you have created the target with `add_executable(e-multicam ${SRC_FILES})`

**Reply by piyushmz** [reply]
Now it's throwing 
`
CMake Error at CMakeLists.txt:29 (target_include_directories):
  target_include_directories called with invalid arguments
`

**Reply by piyushmz** [reply]
I've cleared the cmake error but Still not working 
CMakeLists.txt
```
cmake_minimum_required(VERSION 3.16)
project(e-multicam)

include(GNUInstallDirs)

find_package(iceoryx_platform REQUIRED)
find_package(iceoryx_posh CONFIG REQUIRED)
find_package(iceoryx_hoofs CONFIG REQUIRED)
find_package(iceoryx_binding_c CONFIG REQUIRED)
find_package(PkgConfig REQUIRED)

include(IceoryxPackageHelper)
include(IceoryxPlatform)
include(IceoryxPlatformSettings)

pkg_check_modules(gtk3 REQUIRED IMPORTED_TARGET gtk+-3.0)
pkg_search_module(gstreamer REQUIRED IMPORTED_TARGET gstreamer-1.0>=1.4)
pkg_search_module(gstreamer-sdp REQUIRED IMPORTED_TARGET gstreamer-sdp-1.0>=1.4)
pkg_search_module(gstreamer-app REQUIRED IMPORTED_TARGET gstreamer-app-1.0>=1.4)
pkg_search_module(gstreamer-video REQUIRED IMPORTED_TARGET gstreamer-video-1.0>=1.4)


set(CMAKE_C_COMPILER ${CROSS_COMPILE}gcc)

set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -g -Wall -pedantic -DVERSION=\\\"${VERSION}\\\"")

set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -lpthread")

set(CMAKE_CXX_STANDARD 11)

find_package(PkgConfig)
pkg_check_modules(PKG_CONFIG_MODULES gstreamer-1.0 gstreamer-video-1.0 gstreamer-app-1.0 x11)
# include_directories(${PKG_CONFIG_MODULES_INCLUDE_DIRS})
file(GLOB_RECURSE SRC_FILES "*.c")

add_executable(e-multicam ${SRC_FILES})
target_include_directories(e-multicam
    PUBLIC
    ${PKG_CONFIG_MODULES_INCLUDE_DIRS}
)
target_link_libraries(e-multicam pthread m jpeg ${PKG_CONFIG_MODULES_LDFLAGS})
target_link_libraries(e-multicam 
    iceoryx_binding_c::iceoryx_binding_c
    PkgConfig::gtk3
    PkgConfig::gstreamer
    PkgConfig::gstreamer-sdp
    PkgConfig::gstreamer-app
    PkgConfig::gstreamer-video
    )

install(TARGETS e-multicam DESTINATION ./install)

```

Error is same 

```
$ make -j12
gcc -g -Wall -pedantic -DVERSION=\"1.0\"  `pkg-config --cflags --libs gstreamer-1.0 gstreamer-video-1.0 gstreamer-app-1.0 x11`  -I/usr/include/gstreamer-1.0 -c main.c -o main.o
main.c:9:10: fatal error: iceoryx_binding_c/publisher.h: No such file or directory
    9 | #include "iceoryx_binding_c/publisher.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [Makefile:42: main.o] Error 1

```

**Reply by piyushmz** [reply]
@elfenpiff Thanks I've resolved it with a bit of a workaround so whenever I was doing make following was the error 
```
[100%] Linking CXX executable e-multicam
/usr/bin/ld: CMakeFiles/e-multicam.dir/main.c.o: in function `main':
/home/minuszero/Work/e2/main.c:226: multiple definition of `main'; CMakeFiles/e-multicam.dir/CMakeFiles/3.16.3/CompilerIdC/CMakeCCompilerId.c.o:/home/minuszero/Work/e2/build/CMakeFiles/3.16.3/CompilerIdC/CMakeCCompilerId.c:647: first defined here
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/e-multicam.dir/build.make:241: e-multicam] Error 1
make[2]: Leaving directory '/home/minuszero/Work/e2/build'
make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/e-multicam.dir/all] Error 2
make[1]: Leaving directory '/home/minuszero/Work/e2/build'
make: *** [Makefile:84: all] Error 2
make: Leaving directory '/home/minuszero/Work/e2/build'
```
so I've to do to the `home/minuszero/Work/e2/build/CMakeFiles/3.16.3/CompilerIdC/CMakeCCompilerId.c:647` and delete the function which is as below. I still don't know why it get's generated.
```
#ifdef ID_VOID_MAIN
void main() {}
#else
# if defined(__CLASSIC_C__)
int main(argc, argv) int argc; char *argv[];
# else
int main(int argc, char* argv[])
# endif
{
  int require = 0;
  require += info_compiler[argc];
  require += info_platform[argc];
  require += info_arch[argc];
#ifdef COMPILER_VERSION_MAJOR
  require += info_version[argc];
#endif
#ifdef COMPILER_VERSION_INTERNAL
  require += info_version_internal[argc];
#endif
#ifdef SIMULATE_ID
  require += info_simulate[argc];
#endif
#ifdef SIMULATE_VERSION_MAJOR
  require += info_simulate_version[argc];
#endif
#if defined(__CRAYXE) || defined(__CRAYXC)
  require += info_cray[argc];
#endif
  require += info_language_dialect_default[argc];
  (void)argv;
  return require;
}
#endif
```

**Reply by elfenpiff** [reply]
@piyushmz It seems you encountered some internal problems from a library/binary/framework I am not familiar with. Furthermore the code you presented here is hard to understand and raises more question which I cannot answer. I am not the one who could help you from here on.

Sadly, you are on your own from here. But if you encounter another issue which originates directly in iceoryx feel free to ask again. I am happy to help.

**Reply by piyushmz** [reply]
@elfenpiff  I've got it. Thanks very much you were of great help.
